sources/web/datalab/polymer/components/file-browser/file-browser.ts

```diff
@@ -285,6 +299,49 @@ class FileBrowserElement extends Polymer.Element implements DatalabPageElement {
     this._loadStartupPath(fileId);
   }
 
+  async _fileManagerTypeListChanged() {
+    if (!this.fileManagerTypeList) {
+      const settings = await SettingsManager.getAppSettingsAsync();
+      // Fall back to Jupyter if settings are somehow outdated.
+      const types = settings.supportedFileBrowserSources || ['jupyter'];
+      this.fileManagerTypeList = types.map((source) =>
+          FileManagerFactory.fileManagerNameToType(source));
+    }
+
+    const menu = this.$.fileSourcesDropdown as HTMLDivElement;
+    // Build and add buttons to the switcher dropdown menu
+    Utils.deleteAllChildren(menu);
+
+    this.fileManagerTypeList.forEach((type) => {
+      const config = FILE_MANAGER_CONFIG.get(type);
```

Comment:
This feels kind of like directly accessing the internals of a global. I think I might rather see this encapsulated in a method within FileManagerFactory.

Reply:
Agreed. Done.